src/pybind/mgr/prometheus/module.py

```diff
@@ -961,6 +970,41 @@ def get_osd_blocklisted_entries(self) -> None:
         for stat in OSD_BLOCKLIST:
             self.metrics['cluster_{}'.format(stat)].set(int(blocklist_count))
 
+    @profile_method()
+    def get_mirror_schedule(self) -> None:
+        r = self.mon_command({
+            'prefix': 'rbd mirror snapshot schedule list',
+            'format': 'json'
+        })
+        self.log.info(r)
+        if not r[0]:
+            schedules = json.loads(r[1])
+            for key, value in schedules.items():
+                parts = key.split('/')
+                if len(parts) == 1:  # global pool
+                    pool_name = parts[0]
+                    namespace = ''
+                    image_name = ''
+                elif len(parts) == 2:  # pool and namespace
+                    pool_name, namespace = parts
+                    image_name = ''
+                else:  # pool, namespace, and image
+                    pool_name, namespace, image_name = parts
+                interval = value['schedule'][0]['interval']
+
+                match = re.match(r'^(\d+)(d|h|m)?$', interval)
+                if not match:
+                    raise ValueError("Invalid interval ({})".format(interval))
+                minutes = int(match.group(1))
+                if match.group(2) == 'd':
+                    minutes *= 60 * 24
+                elif match.group(2) == 'h':
+                    minutes *= 60
```

Comment:
This code is carried over from the `Interval` class in `rbd_support`. Importing `rbd_support` module to use the `Interval` class results in failure in adding schedule in a cluster.

To have this metric for all the images instead of just the level spec, would we need to rewrite the logic in LevelSpec?

Reply:
It looks like you are trying to take an output that reports multiple values with different semantics and condense it into a single integer.  Have you considered that multiple schedules (and not necessarily at different levels even) may apply to a single image and that they aren't guaranteed to align?

Consider a case where there is a global schedule for 1h and an image level schedule for 15m.  In this case they align and a common interval can be trivially derived -- 15m.  But, if an image level schedule was set for 25m, these intervals would align only every five hours.  Deriving 25m in that case wouldn't be strictly correct since the expected interval between every second and third snapshot would vary from 5m to 25m.

This is further complicated by the fact that schedules have start times (optional `<start-time>` argument of `rbd mirror snapshot schedule add` command).  One could configure four 1h schedules at :00, :15, :30 and :45 offsets and expect a 15m effective interval while each of the four level specs would have a 1h interval.

I would encourage you to think through all these scenarios before attempting to rewrite anything.

Reply:
@idryomov, in the case you are pointing we will have something like:

rbd_mirror_schedule_interval{"pool_name": "pool1", "namespace": "", "image": ""} = 60
rbd_mirror_schedule_interval{"pool_name": "pool1", "namespace": "", "image": "imageX"} = 25

We are reflecting in the metrics properly the fact that imageX has two schedules. So i think it is not a problem. 

You are right that we will need to add as label the "start date", because that determines when is going to happen the next schedule. So probably is a good idea to add the "start_date" as a label in the metric. What do you think @weirdwiz , @idryomov ?

Note: This information will be used in ODF to know what is the situation of the schedule for a specific image/set of images. And with these new metrics i think we will have the data needed to do the job.

Reply:
> @idryomov, in the case you are pointing we will have something like:
> 
> rbd_mirror_schedule_interval{"pool_name": "pool1", "namespace": "", "image": ""} = 60
> rbd_mirror_schedule_interval{"pool_name": "pool1", "namespace": "", "image": "imageX"} = 25
> 
> We are reflecting in the metrics properly the fact that imageX has two schedules. So i think it is not a problem.

So it appears that what you are aiming to do is simply re-package `rbd mirror snapshot list` output into a set of metrics?  Divyansh's "To have this metric for all the images instead of just the level spec ..." question above made me think that the idea was to report a single interval for each image.

> 
> You are right that we will need to add as label the "start date", because that determines when is going to happen the next schedule. So probably is a good idea to add the "start_date" as a label in the metric. What do you think @weirdwiz , @idryomov ?

Technically, "start date" is part of the schedule (value), not part of the level spec (key).

> 
> Note: This information will be used in ODF to know what is the situation of the schedule for a specific image/set of images. And with these new metrics i think we will have the data needed to do the job.

Then ODF would need to know that `rbd_mirror_schedule_interval{"pool_name": "pool1", "namespace": "", "image": ""}` applies to all images in `pool1` while `rbd_mirror_schedule_interval{"pool_name": "pool1", "namespace": "", "image": "imageX"}` applies only to `imageX` (I'm ignoring namespaces for the moment).  Are you planning to duplicate level spec matching logic in ODF?

Reply:
This data will be used to write alerts, but I don't think we'll be able to write the logic in just prometheus queries. 
I've added a new metric which reports the timestamp of the next snapshot for that image, we could possibly use that to find out the intervals applied to the image. But the problem is that the interval metric contains pool_id and image_id, so it'll be harder to correlate.

Reply:
> This data will be used to write alerts, but I don't think we'll be able to write the logic in just prometheus queries. I've added a new metric which reports the timestamp of the next snapshot for that image, we could possibly use that to find out the intervals applied to the image.

Then I'm not sure I see the point of adding these "metrics" (they aren't really metrics -- `rbd mirror snapshot schedule list` shows what schedules are set up, not how the system is doing, for example).  The alerts should already be written, at least in the form of a draft.  Otherwise how do you know/can proof that what is being added here is sufficient?

> But the problem is that the interval metric contains pool_id and image_id, so it'll be harder to correlate.

Taking this as example, I'm pretty sure that you would eventually need to change one of the metrics to match the other in terms of IDs vs names since attempting to correlate on the side is indeed hard.  Further, with multiple schedules configured, the effective interval (time to next snapshot) can jump around a lot.

I think all of this calls for a demonstration of the whole thing working end-to-end, from a schedule not being met to an alert being raised, so that it's clear what exactly is required at the lowest layer.